packages/connect/src/connector/smoldot-light.ts

```diff
@@ -22,26 +22,68 @@ const getStart = () => {
   return startPromise
 }
 
-let clientNumReferences = 0
+const clientReferences: Config[] = [] // Note that this can't be a set, as the same config is added/removed multiple times
 let clientPromise: Promise<Client> | null = null
-const getClientAndIncRef = (): Promise<Client> => {
+let clientReferencesMaxLogLevel = 3
+const getClientAndIncRef = (config: Config): Promise<Client> => {
+  if (config.maxLogLevel && config.maxLogLevel > clientReferencesMaxLogLevel)
+    clientReferencesMaxLogLevel = config.maxLogLevel
+
   if (clientPromise) {
-    clientNumReferences += 1
+    clientReferences.push(config)
     return clientPromise
   }
 
   clientPromise = getStart().then((start) =>
     start({
       forbidTcp: true, // In order to avoid confusing inconsistencies between browsers and NodeJS, TCP connections are always disabled.
       forbidNonLocalWs: true, // Prevents browsers from emitting warnings if smoldot tried to establish non-secure WebSocket connections
-      maxLogLevel: 4 /* no debug/trace messages */,
+      maxLogLevel: 9999999, // The actual level filtering is done in the logCallback
       cpuRateLimit: 0.5, // Politely limit the CPU usage of the smoldot background worker.
+      logCallback: (level, target, message) => {
+        if (level > clientReferencesMaxLogLevel) return
+
+        // The first parameter of the methods of `console` has some printf-like substitution
+        // capabilities. We don't really need to use this, but not using it means that the logs
+        // might not get printed correctly if they contain `%`.
+        if (level <= 1) {
+          console.error("[%s] %s", target, message)
+        } else if (level === 2) {
+          console.warn("[%s] %s", target, message)
+        } else if (level === 3) {
+          console.info("[%s] %s", target, message)
+        } else if (level === 4) {
+          console.debug("[%s] %s", target, message)
+        } else {
+          console.trace("[%s] %s", target, message)
+        }
+      },
     }),
   )
-  clientNumReferences += 1
+  clientReferences.push(config)
   return clientPromise
 }
 
+// Must be passed the exact same object as was passed to {getClientAndIncRef}
+const decRef = (config: Config) => {
+  const idx = clientReferences.indexOf(config)
+  if (idx === -1) throw new Error("Internal error within smoldot-light")
+  clientReferences.splice(idx, 1)
+
+  // Update `clientReferencesMaxLogLevel`
+  // Note how it is set back to 3 if there is no reference anymore
+  clientReferencesMaxLogLevel = 3
+  for (const cfg of clientReferences.values()) {
+    if (cfg.maxLogLevel && cfg.maxLogLevel > clientReferencesMaxLogLevel)
```

Comment:
I think this is the same:
```suggestion
    if (cfg?.maxLogLevel > clientReferencesMaxLogLevel)
```